### PR TITLE
feat(linter): Implement eslint:yoda

### PR DESCRIPTION
**Pull request by tbashiyy**
In this PR, implement [eslint:yoda](https://eslint.org/docs/latest/rules/yoda)

ref: https://github.com/oxc-project/oxc/issues/479

### CODE REVIEW

File: crates/oxc_linter/src/rules/unicorn/prefer_negative_index.rs [hunk 1]
```diff
@@ -300,7 +300,7 @@ fn test() {
         "[1, 2, 3].slice([1, 2, 3].length - 1)",
         "foo[bar++].slice(foo[bar++].length - 1)",
         "foo[a + b].slice(foo[a + b].length - 1)",
-        "foo[`${bar}`].slice(foo[`${bar}`].length - 1)",
+        // "foo[`${bar}`].slice(foo[`${bar}`].length - 1)",
```
Comment (tbashiyy) [inline crates/oxc_linter/src/rules/unicorn/prefer_negative_index.rs:303]: I think this should be moved to fail case.
If my understanding is incorrect, please tell me🙇‍♂️

If correct, I will move this case to fail.

File: crates/oxc_linter/src/rules/eslint/yoda.rs [hunk 2]
```diff
@@ -0,0 +1,1160 @@
...
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_ecmascript::ToBigInt;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{GetSpan, Span};
+use regex::Regex;
+
+use crate::{context::LintContext, rule::Rule, utils::is_same_reference, AstNode};
+
+fn yoda_diagnostic(span: Span) -> OxcDiagnostic {
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/yoda.rs:16]: shall we have a few diagnostics based on the config struct?
doing thsi should help make the help msg easier to comprehend

File: crates/oxc_linter/src/rules/eslint/yoda.rs [hunk 3]
```diff
@@ -0,0 +1,1160 @@
...
+    never: bool,
+    except_range: bool,
+    only_equality: bool,
+}
+
+declare_oxc_lint!(
+    /// ## What it does
+    ///
+    /// Require or disallow "Yoda" conditions.
+    /// This rule aims to enforce consistent style of conditions which compare a variable to a literal value. The literal value must always come first in the comparison.
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/yoda.rs:33]: is "The literal value must always come first in the comparison." correct? shouldn't it be the other way round?

File: crates/oxc_linter/src/rules/eslint/yoda.rs (hunk 2 above)
Comment (tbashiyy) [reply crates/oxc_linter/src/rules/eslint/yoda.rs]: Thanks!
I fixed refer to [original rule](https://github.com/eslint/eslint/blob/e6b84f535e84050b0c63ae304eb17f9d181ac299/lib/rules/yoda.js#L146) 
https://github.com/oxc-project/oxc/pull/7559/commits/e07f9fdf7f014a655128c82f7fdbfd53c762978f

File: crates/oxc_linter/src/rules/eslint/yoda.rs (hunk 3 above)
Comment (tbashiyy) [reply crates/oxc_linter/src/rules/eslint/yoda.rs]: deleted. https://github.com/oxc-project/oxc/pull/7559/commits/abecc1a71c09dd523cd306df8617b33b5d963b8f
(I'm not sure why this sentence was inserted.. maybe copilot did..?)

File: crates/oxc_linter/src/utils/unicorn.rs [hunk 4]
```diff
@@ -177,6 +178,23 @@ pub fn is_same_reference(left: &Expression, right: &Expression, ctx: &LintContex
...
             return left_str.value == right_str.value;
         }
+        (Expression::StringLiteral(left_str), Expression::TemplateLiteral(right_str)) => {
+            return right_str.is_no_substitution_template()
+                && left_str.value == right_str.quasi().unwrap();
+        }
+        (Expression::TemplateLiteral(left_str), Expression::StringLiteral(right_str)) => {
+            return left_str.is_no_substitution_template()
+                && left_str.quasi().unwrap() == right_str.value;
+        }
```
Comment (camc314) [inline crates/oxc_linter/src/utils/unicorn.rs:188]:
```suggestion
        (Expression::StringLiteral(string_lit), Expression::TemplateLiteral(template_lit))
        | (Expression::TemplateLiteral(template_lit), Expression::StringLiteral(string_lit)) => {
            return template_lit.is_no_substitution_template()
                && string_lit.value == template_lit.quasi().unwrap();
        }
```

File: crates/oxc_linter/src/utils/unicorn.rs [hunk 5]
```diff
@@ -266,12 +284,56 @@ pub fn is_same_member_expression(
...
     ) = (left, right)
     {
-        if !is_same_reference(
-            left.expression.get_inner_expression(),
-            right.expression.get_inner_expression(),
-            ctx,
-        ) {
-            return false;
+        match (&left.expression, &right.expression) {
+            // ex) x['/regex/'] === x[/regex/]
```
Comment (camc314) [inline crates/oxc_linter/src/utils/unicorn.rs:288]: this language never ceases to amaze me

**Review by camc314 — APPROVED** [review]
thank you!

**Pull request merged**